contrib/platform/src/com/sun/jna/platform/win32/PsapiUtil.java

```diff
@@ -0,0 +1,118 @@
+ * You may obtain a copy of the Apache License at:
+ *
+ * http://www.apache.org/licenses/
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "AL2.0".
+ */
+package com.sun.jna.platform.win32;
+
+import java.util.Arrays;
+
+import com.sun.jna.Memory;
+import com.sun.jna.Native;
+import com.sun.jna.platform.win32.WinDef.DWORD;
+import com.sun.jna.platform.win32.WinNT.HANDLE;
+import com.sun.jna.ptr.IntByReference;
+
+/**
+ * Psapi utility API.
+ *
+ * @author Torbj&ouml;rn Svensson, azoff[at]svenskalinuxforeninen.se
+ */
+public abstract class PsapiUtil {
+
+    /**
+     * Retrieves the process identifier for each process object in the system.
+     *
+     * @return Array of pids
+     */
+    public static int[] enumProcesses() {
+        int size = 0;
+        int[] lpidProcess = null;
+        IntByReference lpcbNeeded = new IntByReference();
+        do {
+            size = size + 1024;
+            lpidProcess = new int[size];
+            if (!Psapi.INSTANCE.EnumProcesses(lpidProcess, size * DWORD.SIZE, lpcbNeeded)) {
+                throw new Win32Exception(Native.getLastError());
+            }
+        } while (size == lpcbNeeded.getValue() / DWORD.SIZE);
+
+        return Arrays.copyOf(lpidProcess, lpcbNeeded.getValue() / DWORD.SIZE);
+    }
+
+    /**
+     * Retrieves the name of the process identified by <i>pid</i>
+     *
+     * @param pid The process identifier
+     * @return The name of the process
+     */
+    public static String getProcessName(int pid) {
+        HANDLE hProcess = null;
+        Win32Exception we = null;
+
+        try {
+            hProcess = Kernel32.INSTANCE.OpenProcess(WinNT.PROCESS_QUERY_INFORMATION | WinNT.PROCESS_VM_READ, false, pid);
+            if (hProcess == null) {
+                throw new Win32Exception(Native.getLastError());
+            }
+
+            return getProcessName(hProcess);
+        } catch (Win32Exception e) {
+            we = e;
+            throw we; // re-throw to invoke finally block
+        } finally {
+            try {
+                Kernel32Util.closeHandle(hProcess);
+            } catch (Win32Exception e) {
+                if (we == null) {
+                    we = e;
+                } else {
+                    we.addSuppressed(e);
+                }
+            }
+            if (we != null) {
+                throw we;
+            }
+        }
+    }
+
+    /**
+     * Retrieves the name of the process handle
+     *
+     * @param hProcess The process handle
+     * @return The name of the process
+     */
+    public static String getProcessName(HANDLE hProcess) {
+        int maxStringLength = 0;
+        int length = 0;
+        Memory lpFilename = null;
+        do {
+            maxStringLength = maxStringLength + 256;
+            lpFilename = new Memory(maxStringLength * Native.WCHAR_SIZE);
+            length = Psapi.INSTANCE.GetModuleFileNameEx(hProcess, null, lpFilename, maxStringLength);
```

Comment:
I'll remove it as there is another function that is more reliable. Thanks for the tip!

Reply:
I have one question regarding the QueryFullProcessImageName() implementation; can't the path be longer than 260 if using UNC?

Reply:
Good point.  It can also be exceeded if a user edits their registry with `LongPathsEnabled` (added in the Win10 Anniversary Update).  So this code should probably check the length on return and increase the buffer size.